# GROUP 4: MEMBERS
1.KALEBO
2.JACK
3.MUSA
4.ERICK
5.JANETH
6.ADAM




# PART A: PANDAS DATAFRAME
 


## A1: DATA FRAME FROM A LIST


In [1]:
import pandas as pd

In [2]:


#  A list of 5 records with 3 fields each:( ID, Name, Age)
data = [
    [101, "Janet", 20],
    [102, "Erick", 20],
    [103, "Adam", 18],
    [104, "Musa", 20],
    [105, "Jack", 19]
]

# Converting DataFrame and assigning column names
df = pd.DataFrame(data, columns=["ID", "Name", "Age"])

#  Displaying DataFrame
print(df)

    ID   Name  Age
0  101  Janet   20
1  102  Erick   20
2  103   Adam   18
3  104   Musa   20
4  105   Jack   19


The code above involves converting a list made up of several series into pandas dataFrame
using function (pd.DataFrame) ,print


##  A2: DATA FRAME FROM A DICTIONARY


In [4]:

# dictionary of 3 columns and 5records
data = {
    'Student_ID': [101, 102, 103, 104, 105],
    'Name': ['Janet', 'Erick', 'Adam', 'Musa', 'Jack'],
    'Score': [85.5, 92.0, 78.0, 88.5, 95.0]
}

# Converting dictionary into Pandas DataFrame
df = pd.DataFrame(data)

# Displaying DataFrame
print("DataFrame")
print(df)

# Displaying columns and data types
print("\nColumns")
print(df.columns)

print("\nData Types ")
print(df.dtypes)

DataFrame
   Student_ID   Name  Score
0         101  Janet   85.5
1         102  Erick   92.0
2         103   Adam   78.0
3         104   Musa   88.5
4         105   Jack   95.0

Columns
Index(['Student_ID', 'Name', 'Score'], dtype='str')

Data Types 
Student_ID      int64
Name              str
Score         float64
dtype: object


The code above involves converting a dictionary of 3 columns and 5 records into a DataFrame  .using the function pd.DataFrame ,print ,df.columns ,df.types ,


# A3: DATA FRAME FROM  SERIES


In [58]:
# three separate Pandas Series
names = pd.Series(['Janet', 'Erick', 'Adam', 'Musa', 'Jack'])
ages = pd.Series([20, 20, 18, 20, 19])
scores = pd.Series([85.5, 92.0, 78.0, 88.5, 95.0])

# Combining them into a DataFrame
df = pd.DataFrame({'Student Name': names, 'Age': ages,'Score': scores})

# Displaying  DataFrame
print(df)

  Student Name  Age  Score
0        Janet   20   85.5
1        Erick   20   92.0
2         Adam   18   78.0
3         Musa   20   88.5
4         Jack   19   95.0


The code above involves converting  several series into pandas dataFrame
using function (pd.DataFrame) ,print

  RELATIONSHIP BETWEEN A SERIES AND A DATAFRAME
Series:  is a 1-dimensional labeled array capable of holding data of any single type (integer, float, string)

DataFrame: is a 2-dimensional tabular structure with rows and columns.(is a combination of several series)


# PART 2: KAGGLE DATA ANALYSIS

#### link of the weather dataset
path = kagglehub.dataset_download("muthuj7/weather-dataset")



## 1: IMPORTING


In [12]:

# Reading CSV file
df = pd.read_csv('weatherHistory.csv')

reading the information from the CSV file using the  function  df = pd.read_csv

In [ ]:
# First 10 rows
df.head(10)


,Formatted Date,Summary,Precip Type,Temperature (C),Apparent Temperature (C),Humidity,Wind Speed (km/h),Wind Bearing (degrees),Visibility (km),Loud Cover,Pressure (millibars),Daily Summary
0,2006-04-01 00:00:00.000 +0200,Partly Cloudy,rain,9.472222,7.388889,0.89,14.1197,251.0,15.8263,0.0,1015.13,Partly cloudy throughout the day.
1,2006-04-01 01:00:00.000 +0200,Partly Cloudy,rain,9.355556,7.227778,0.86,14.2646,259.0,15.8263,0.0,1015.63,Partly cloudy throughout the day.
2,2006-04-01 02:00:00.000 +0200,Mostly Cloudy,rain,9.377778,9.377778,0.89,3.9284,204.0,14.9569,0.0,1015.94,Partly cloudy throughout the day.
3,2006-04-01 03:00:00.000 +0200,Partly Cloudy,rain,8.288889,5.944444,0.83,14.1036,269.0,15.8263,0.0,1016.41,Partly cloudy throughout the day.
4,2006-04-01 04:00:00.000 +0200,Mostly Cloudy,rain,8.755556,6.977778,0.83,11.0446,259.0,15.8263,0.0,1016.51,Partly cloudy throughout the day.
5,2006-04-01 05:00:00.000 +0200,Partly Cloudy,rain,9.222222,7.111111,0.85,13.9587,258.0,14.9569,0.0,1016.66,Partly cloudy throughout the day.
6,2006-04-01 06:00:00.000 +0200,Partly Cloudy,rain,7.733333,5.522222,0.95,12.3648,259.0,9.9820,0.0,1016.72,Partly cloudy throughout the day.
7,2006-04-01 07:00:00.000 +0200,Partly Cloudy,rain,8.772222,6.527778,0.89,14.1519,260.0,9.9820,0.0,1016.84,Partly cloudy throughout the day.
8,2006-04-01 08:00:00.000 +0200,Partly Cloudy,rain,10.822222,10.822222,0.82,11.3183,259.0,9.9820,0.0,1017.37,Partly cloudy throughout the day.
9,2006-04-01 09:00:00.000 +0200,Partly Cloudy,rain,13.772222,13.772222,0.72,12.5258,279.0,9.9820,0.0,1017.22,Partly cloudy throughout the day.


Reading the first 10 items from CSV file using the function  df.head(here you input the required no. you are reading)

In [6]:
# Last 5 rows
df.tail(5)


,Formatted Date,Summary,Precip Type,Temperature (C),Apparent Temperature (C),Humidity,Wind Speed (km/h),Wind Bearing (degrees),Visibility (km),Loud Cover,Pressure (millibars),Daily Summary
96448,2016-09-09 19:00:00.000 +0200,Partly Cloudy,rain,26.016667,26.016667,0.43,10.9963,31.0,16.1000,0.0,1014.36,Partly cloudy starting in the morning.
96449,2016-09-09 20:00:00.000 +0200,Partly Cloudy,rain,24.583333,24.583333,0.48,10.0947,20.0,15.5526,0.0,1015.16,Partly cloudy starting in the morning.
96450,2016-09-09 21:00:00.000 +0200,Partly Cloudy,rain,22.038889,22.038889,0.56,8.9838,30.0,16.1000,0.0,1015.66,Partly cloudy starting in the morning.
96451,2016-09-09 22:00:00.000 +0200,Partly Cloudy,rain,21.522222,21.522222,0.60,10.5294,20.0,16.1000,0.0,1015.95,Partly cloudy starting in the morning.
96452,2016-09-09 23:00:00.000 +0200,Partly Cloudy,rain,20.438889,20.438889,0.61,5.8765,39.0,15.5204,0.0,1016.16,Partly cloudy starting in the morning.


Reading the last 5 items from CSV file using the function  df.tail(here you input the required no. you are reading)

In [47]:
# Shape (number of rows, number of columns)
df.shape


(96453, 15)

Identifying the number of rows and the number of columns from CSV file using the function  df.shape()

In [67]:
# Columns (list of all column names)
df = pd.read_csv('weatherHistory.csv')
df.columns

Index(['Formatted Date', 'Summary', 'Precip Type', 'Temperature (C)',
       'Apparent Temperature (C)', 'Humidity', 'Wind Speed (km/h)',
       'Wind Bearing (degrees)', 'Visibility (km)', 'Loud Cover',
       'Pressure (millibars)', 'Daily Summary'],
      dtype='str')

This involves reading or retrieving of all the column names from the CSV file and returns an index object containing all labels

In [49]:
# Data types of each column
df.dtypes

Formatted Date              datetime64[us, UTC]
Summary                                     str
Precip Type                                 str
Temperature (C)                         float64
Apparent Temperature (C)                float64
Humidity                                float64
Wind Speed (km/h)                       float64
Wind Bearing (degrees)                  float64
Visibility (km)                         float64
Loud Cover                              float64
Pressure (millibars)                    float64
Daily Summary                               str
Month                                     int32
Hour                                      int32
Day                                       int32
dtype: object

To read the data types from CSV file using the function df.dtypes that returns a Series containing the data type (dtype) of every column in the DataFrame.


## 2: STATISTICS


In [54]:
df['Temperature (C)'].mean()

np.float64(11.93267843751188)

finding the mean or average of temperature from the csv file using the function df.mean()

In [ ]:
df['Temperature (C)'].max()

np.float64(39.90555555555555)

finding the maximum value of temperature from the csv file using the function df.max()

In [ ]:
# minimum temp
df['Temperature (C)'].min()

np.float64(-21.822222222222223)

finding the minimum  value of temperature from the csv file using the function df.min()

In [ ]:
# average humidity
df['Humidity'].mean()

np.float64(0.7348989663359357)

finding the average of Humidity from the csv file using the function df.mean() which is = num py double precision floating number point

In [ ]:
# Average wind speed 
df['Wind Speed (km/h)'].mean()

np.float64(10.810640140793963)

finding the average of wind speed from the csv file using the function df.mean()

In [57]:
df['Temperature (C)'].std()

np.float64(9.551546320657026)

The standard deviation from the CSV data file temperature is = 9.551546 which is in num py double precision floating number of 15 decimal places

In [43]:
df.describe()

,Temperature (C),Apparent Temperature (C),Humidity,Wind Speed (km/h),Wind Bearing (degrees),Visibility (km),Loud Cover,Pressure (millibars),Month,Hour,Day
count,96453.000000,96453.000000,96453.000000,96453.000000,96453.000000,96453.000000,96453.0,96453.000000,96453.000000,96453.000000,96453.000000
mean,11.932678,10.855029,0.734899,10.810640,187.509232,10.347325,0.0,1003.235956,6.523001,11.499891,15.727100
std,9.551546,10.696847,0.195473,6.913571,107.383428,4.192123,0.0,116.969906,3.448501,6.922049,8.802216
min,-21.822222,-27.716667,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,1.000000,0.000000,1.000000
25%,4.688889,2.311111,0.600000,5.828200,116.000000,8.339800,0.0,1011.900000,4.000000,6.000000,8.000000
50%,12.000000,12.000000,0.780000,9.965900,180.000000,10.046400,0.0,1016.450000,7.000000,11.000000,16.000000
75%,18.838889,18.838889,0.890000,14.135800,290.000000,14.812000,0.0,1021.090000,10.000000,17.000000,23.000000
max,39.905556,39.344444,1.000000,63.852600,359.000000,16.100000,0.0,1046.380000,12.000000,23.000000,31.000000


The function describe gives and displays the basic statistics data of number of columns,the mean, standard deviation,minimum data,the 25%,50% percentiles or meadian of each data column in the CSV data read also the maximum value of each data column



### 3: CLEANING


In [19]:
# 1. Check missing values
print("--- Missing Values Count ---")
print(df.isnull().sum())

--- Missing Values Count ---
Formatted Date                0
Summary                       0
Precip Type                 517
Temperature (C)               0
Apparent Temperature (C)      0
Humidity                      0
Wind Speed (km/h)             0
Wind Bearing (degrees)        0
Visibility (km)               0
Loud Cover                    0
Pressure (millibars)          0
Daily Summary                 0
Month                         0
dtype: int64


The function determines the empty data rows and columns in the overall CSV file

In [20]:
# 2. Check duplicates
print("\n--- Duplicate Rows Count ---")
print(df.duplicated().sum())


--- Duplicate Rows Count ---
24


This counts the total number of repeated(duplicated) rows in the DataFramres and gives True if the row is a duplicate of the previous row  and gives False if the rows are all unique 
and for True the value counts for specific interger values and if False the sum is zero(0)

In [23]:
# Convert Formatted Date column
df['Formatted Date'] = pd.to_datetime(df['Formatted Date'], utc=True)

# Inspect converted output and data type
print(df['Formatted Date'].head(26))
print("\nData Type:", df['Formatted Date'].dtype)

0    2006-03-31 22:00:00+00:00
1    2006-03-31 23:00:00+00:00
2    2006-04-01 00:00:00+00:00
3    2006-04-01 01:00:00+00:00
4    2006-04-01 02:00:00+00:00
5    2006-04-01 03:00:00+00:00
6    2006-04-01 04:00:00+00:00
7    2006-04-01 05:00:00+00:00
8    2006-04-01 06:00:00+00:00
9    2006-04-01 07:00:00+00:00
10   2006-04-01 08:00:00+00:00
11   2006-04-01 09:00:00+00:00
12   2006-04-01 10:00:00+00:00
13   2006-04-01 11:00:00+00:00
14   2006-04-01 12:00:00+00:00
15   2006-04-01 13:00:00+00:00
16   2006-04-01 14:00:00+00:00
17   2006-04-01 15:00:00+00:00
18   2006-04-01 16:00:00+00:00
19   2006-04-01 17:00:00+00:00
20   2006-04-01 18:00:00+00:00
21   2006-04-01 19:00:00+00:00
22   2006-04-01 20:00:00+00:00
23   2006-04-01 21:00:00+00:00
24   2006-04-09 22:00:00+00:00
25   2006-04-09 23:00:00+00:00
Name: Formatted Date, dtype: datetime64[us, UTC]

Data Type: datetime64[us, UTC]


The data selected in the first 26 columns is formated and changed into the specific format and also displays the data type of the selected column

In [24]:
# 3. Extract Hour, Day, and Month
df['Hour'] = df['Formatted Date'].dt.hour
df['Day'] = df['Formatted Date'].dt.day
df['Month'] = df['Formatted Date'].dt.month

# 4. Display the results
print(df[['Formatted Date', 'Hour', 'Day', 'Month']].head(5))

             Formatted Date  Hour  Day  Month
0 2006-03-31 22:00:00+00:00    22   31      3
1 2006-03-31 23:00:00+00:00    23   31      3
2 2006-04-01 00:00:00+00:00     0    1      4
3 2006-04-01 01:00:00+00:00     1    1      4
4 2006-04-01 02:00:00+00:00     2    1      4


After the data was formated above the hour,day,month are extracted a simple of the first 5 columns

In [ ]:
# Identify all numerical columns in the DataFrame
num_cols = df.select_dtypes(include=['number']).columns

# Check the count of missing values per numerical column
print("Missing values per numerical column:")
print(df[num_cols].isnull().sum())

df[num_cols] = df[num_cols].fillna(df[num_cols].mean())
print("\nRemaining missing values in numerical columns:")
print(df[num_cols].isnull().sum())

Missing values per numerical column:
Temperature (C)             0
Apparent Temperature (C)    0
Humidity                    0
Wind Speed (km/h)           0
Wind Bearing (degrees)      0
Visibility (km)             0
Loud Cover                  0
Pressure (millibars)        0
Month                       0
Hour                        0
Day                         0
dtype: int64

Remaining missing values in numerical columns:
Temperature (C)             0
Apparent Temperature (C)    0
Humidity                    0
Wind Speed (km/h)           0
Wind Bearing (degrees)      0
Visibility (km)             0
Loud Cover                  0
Pressure (millibars)        0
Month                       0
Hour                        0
Day                         0
dtype: int64


The result displayed is the answer of the missing values handled by the method of sum and mean and also with their respective data types

### 4: SELECTING

In [ ]:
# Records temperature above average
avg_temp = df['Temperature (C)'].mean()
above_avg_temp = df[df['Temperature (C)'] > avg_temp]
print("Records with Temperature Above Average")
print(above_avg_temp.head())

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv("your_dataset.csv")

# Calculate average temperature
avg_temp = df['Temperature (C)'].mean()

# Records with temperature above average
above_avg_temp = df[df['Temperature (C)'] > avg_temp]

print("Average Temperature:", avg_temp)
print("Records with Temperature Above Average")
print(above_avg_temp.head())

In [ ]:
# • Records with high humidity (e.g., Humidity > 0.8 / 80%)
high_humidity = df[df['Humidity'] > 0.8]
print("\nRecords with High Humidity (> 80%) ")
print(high_humidity.head())


--- Records with High Humidity (> 80%) ---
             Formatted Date        Summary Precip Type  Temperature (C)  \
0 2006-03-31 22:00:00+00:00  Partly Cloudy        rain         9.472222   
1 2006-03-31 23:00:00+00:00  Partly Cloudy        rain         9.355556   
2 2006-04-01 00:00:00+00:00  Mostly Cloudy        rain         9.377778   
3 2006-04-01 01:00:00+00:00  Partly Cloudy        rain         8.288889   
4 2006-04-01 02:00:00+00:00  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0          

In [ ]:
# Filtering where 'Precip Type' is recorded (not null / rain or snow):
precip_records = df[df['Precip Type'].notnull()]
print("\nRecords with Active Precipitation ")
print(precip_records.head())


--- Records with Active Precipitation ---
             Formatted Date        Summary Precip Type  Temperature (C)  \
0 2006-03-31 22:00:00+00:00  Partly Cloudy        rain         9.472222   
1 2006-03-31 23:00:00+00:00  Partly Cloudy        rain         9.355556   
2 2006-04-01 00:00:00+00:00  Mostly Cloudy        rain         9.377778   
3 2006-04-01 01:00:00+00:00  Partly Cloudy        rain         8.288889   
4 2006-04-01 02:00:00+00:00  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0           

In [ ]:
# •Selected columns
selected_cols = df[['Formatted Date', 'Summary', 'Temperature (C)', 'Humidity', 'Wind Speed (km/h)']]
print("\nSelected Columns ")
print(selected_cols.head())


--- Selected Columns ---
             Formatted Date        Summary  Temperature (C)  Humidity  \
0 2006-03-31 22:00:00+00:00  Partly Cloudy         9.472222      0.89   
1 2006-03-31 23:00:00+00:00  Partly Cloudy         9.355556      0.86   
2 2006-04-01 00:00:00+00:00  Mostly Cloudy         9.377778      0.89   
3 2006-04-01 01:00:00+00:00  Partly Cloudy         8.288889      0.83   
4 2006-04-01 02:00:00+00:00  Mostly Cloudy         8.755556      0.83   

   Wind Speed (km/h)  
0            14.1197  
1            14.2646  
2             3.9284  
3            14.1036  
4            11.0446  



### 5: GROUPUNG


In [ ]:
 avg_temp_by_month = df.groupby('Month')['Temperature (C)'].mean()
print("\nAverage Temperature by Month ")
print(avg_temp_by_month)


--- Average Temperature by Month ---
Month
1      0.815678
2      2.166880
3      6.914976
4     12.769200
5     16.874455
6     20.728510
7     22.965462
8     22.339134
9     17.503458
10    11.334119
11     6.587549
12     1.625086
Name: Temperature (C), dtype: float64


In [ ]:
#  Average humidity by month
avg_hum_by_month = df.groupby('Month')['Humidity'].mean()
print("\nAverage Humidity by Month")
print(avg_hum_by_month)


--- Average Humidity by Month ---
Month
1     0.850698
2     0.813407
3     0.702849
4     0.641051
5     0.691650
6     0.686418
7     0.639593
8     0.635300
9     0.689163
10    0.774443
11    0.827917
12    0.870398
Name: Humidity, dtype: float64


In [45]:
# Maximum temperature by day
max_temp_by_day = df.groupby(['Month', 'Day'])['Temperature (C)'].max()
print("\nMaximum Temperature by Day")
print(max_temp_by_day.head(10))


Maximum Temperature by Day
Month  Day
1      1      10.000000
       2       8.888889
       3      10.000000
       4      10.966667
       5      11.244444
       6      10.611111
       7      10.116667
       8      10.933333
       9      12.894444
       10     15.738889
Name: Temperature (C), dtype: float64


In [ ]:
# Average wind speed by month
avg_wind_by_month = df.groupby('Month')['Wind Speed (km/h)'].mean()
print("\n--- Average Wind Speed by Month ---")
print(avg_wind_by_month)


--- Average Wind Speed by Month ---
Month
1     11.515865
2     12.180815
3     13.415825
4     11.876139
5     10.953186
6      9.632453
7      9.637124
8      8.942707
9      9.610204
10    10.008495
11    10.947953
12    11.094578
Name: Wind Speed (km/h), dtype: float64



###  6 : COMBINING


In [8]:

#Load and prepare main weather DataFrame
df = pd.read_csv('weatherHistory.csv')
df['Formatted Date'] = pd.to_datetime(df['Formatted Date'], utc=True)
df['Month'] = df['Formatted Date'].dt.month

# Aggregate monthly averages from the main dataset
monthly_weather = df.groupby('Month').agg({
    'Temperature (C)': 'mean',
    'Humidity': 'mean',
    'Wind Speed (km/h)': 'mean'
}).reset_index()

#  Create a second DataFrame containing monthly weather categories
season_data = {
    'Month': list(range(1, 13)),
    'Season': [
        'Winter', 'Winter', 'Spring', 'Spring', 'Spring', 'Summer',
        'Summer', 'Summer', 'Autumn', 'Autumn', 'Autumn', 'Winter'
    ],
    'Category': [
        'Cold/Dry', 'Cold/Dry', 'Mild', 'Mild/Rainy', 'Warm', 'Hot',
        'Hot', 'Hot/Dry', 'Warm', 'Cool/Rainy', 'Cool', 'Cold'
    ]
}
category_df = pd.DataFrame(season_data)

# Merge the two DataFrames on the 'Month' column
merged_df = pd.merge(monthly_weather, category_df, on='Month', how='inner')

# Display result
print("--- Merged Monthly Weather DataFrame ---")
print(merged_df)

--- Merged Monthly Weather DataFrame ---
    Month  Temperature (C)  Humidity  Wind Speed (km/h)  Season    Category
0       1         0.815678  0.850698          11.515865  Winter    Cold/Dry
1       2         2.166880  0.813407          12.180815  Winter    Cold/Dry
2       3         6.914976  0.702849          13.415825  Spring        Mild
3       4        12.769200  0.641051          11.876139  Spring  Mild/Rainy
4       5        16.874455  0.691650          10.953186  Spring        Warm
5       6        20.728510  0.686418           9.632453  Summer         Hot
6       7        22.965462  0.639593           9.637124  Summer         Hot
7       8        22.339134  0.635300           8.942707  Summer     Hot/Dry
8       9        17.503458  0.689163           9.610204  Autumn        Warm
9      10        11.334119  0.774443          10.008495  Autumn  Cool/Rainy
10     11         6.587549  0.827917          10.947953  Autumn        Cool
11     12         1.625086  0.870398          1